src/diagnostic/getnetworkhistory.cpp

```diff
@@ -0,0 +1,145 @@
+ * https://www.gnu.org/licenses/gpl-3.0.en.html) ('GPLv3'). The programs that
+ * are governed by GPLv3.0 are those programs that are located in the folders
+ * src/depends and tests/depends and which include a reference to GPLv3 in their
+ * program files.
+ */
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "libPersistence/BlockStorage.h"
+
+std::string getCsvHeader(uint64_t start, uint64_t stop) {
+  std::string ret = "Node";
+
+  for (uint64_t i = start; i <= stop; ++i) {
+    ret += ",DS epoch " + std::to_string(i);
+  }
+
+  return ret;
+}
+
+void processShards(
+    const DequeOfShard& shards,
+    std::map<std::string, std::map<uint64_t, std::string>>& results,
+    uint64_t dsEpochNo) {
+  uint64_t shardIndex = 0;
+
+  for (auto shardItr = shards.begin(); shardItr != shards.end();
+       ++shardItr, ++shardIndex) {
+    for (size_t peerIndex = 0; peerIndex < shardItr->size(); ++peerIndex) {
+      // get the peer and convert to ip.
+      Peer peer = std::get<1>((*shardItr)[peerIndex]);
+      std::string ip = peer.GetPrintableIPAddress();
+
+      if (results.find(ip) == results.end()) {
+        results[ip] = std::map<uint64_t, std::string>();
+      }
+
+      results[ip][dsEpochNo] = "Shard " + std::to_string(shardIndex) +
+                               " Index " + std::to_string(peerIndex);
+    }
+  }
+}
+
+void processDSCommittee(
+    const DequeOfDSNode& dsCommittee,
+    std::map<std::string, std::map<uint64_t, std::string>>& results,
+    uint64_t dsEpochNo) {
+  uint64_t dsCommitteeIndex = 0;
+
+  for (auto peerItr = dsCommittee.begin(); peerItr != dsCommittee.end();
+       ++peerItr, ++dsCommitteeIndex) {
+    // get the peer and convert to ip.
+    Peer peer = std::get<1>(*peerItr);
+    std::string ip = peer.GetPrintableIPAddress();
+
+    if (results.find(ip) == results.end()) {
+      results[ip] = std::map<uint64_t, std::string>();
+    }
+
+    results[ip][dsEpochNo] = "DS Index " + std::to_string(dsCommitteeIndex);
+  }
+}
+
+void processResults(
+    std::map<std::string, std::map<uint64_t, std::string>>& results,
+    std::string& output, uint64_t blockStart, uint64_t blockStop) {
+  output = getCsvHeader(blockStart, blockStop) + "\n";
+
+  for (auto const& it : results) {
+    std::string ip = it.first;
+
+    std::string row = ip + ",";
+    for (uint64_t block = blockStart; block < blockStop; ++block) {
+      if (results[ip].find(block) != results[ip].end()) {
+        row += results[ip][block] + ",";
+      } else {
+        row += "Not sharded,";
+      }
+    }
+    // Last block.
+    if (results[ip].find(blockStop) != results[ip].end()) {
+      row += results[ip][blockStop];
+    } else {
+      row += "Not sharded";
+    }
+
+    output += row + "\n";
+  }
+}
+
+int main(int argc, char** argv) {
+  if (argc < 2) {
+    std::cout << "[USAGE] " << argv[0]
+              << " <output csv filename> <epoch width required>" << std::endl;
```

Comment:
I think `<epoch width required>` is unused and can be removed?

Reply:
I think it would be good to include the location of the database as a parameter

Reply:
Done!